{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea04b0af",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "\"\"\"\n",
    "Construct heatmap of different gene families\n",
    "\"\"\"\n",
    "import sys\n",
    "sys.path.insert(1, 'workflow/scripts/')\n",
    "sys.path.insert(2, '../../workflow/scripts/')\n",
    "import rnaseqpoptools as rnaseqpop\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from matplotlib.collections import LineCollection\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe571f38",
   "metadata": {
    "tags": [
     "parameters",
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "# Read in parameters from snakemake\n",
    "metadata_path = \"../../config/samples.tsv\"\n",
    "config_path = \"../../config/config.yaml\"\n",
    "pfam_path = \"../../resources/Anogam_long.pep_Pfamscan.seqs.gz\"\n",
    "go_path = \"../../resources/Anogam_long.pep_eggnog_diamond.emapper.annotations.tsv.gz\"\n",
    "normcounts_path = \"../../results/counts/rawcounts.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3e68685",
   "metadata": {},
   "source": [
    "# Gene family heatmaps \n",
    "\n",
    "In this notebook, we summarise gene expression across different protein families (based on PFAM domains) and on GO terms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dc20146",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import yaml\n",
    "with open(config_path) as params_file:\n",
    "    config_params = yaml.safe_load(params_file)\n",
    "\n",
    "comparisons = config_params[\"contrasts\"]\n",
    "comparisons = pd.DataFrame(comparisons, columns=['contrast'])\n",
    "metadata = rnaseqpop.load_metadata(metadata_path)\n",
    "\n",
    "# Read in .csv file containing selection signals and associated metadata\n",
    "pfam_df = pd.read_csv(pfam_path, sep=\"\\s+\", header=None)\n",
    "go_df = pd.read_csv(go_path, sep=\"\\t\", header=None)\n",
    "\n",
    "pfam_df.columns = [\"transcript\", \"pstart\", \"pend\", \"pfamid\", \"domain\", \"domseq\"]\n",
    "go_df.columns = ['transcript', 'GO_terms']\n",
    "\n",
    "gene_annot_df = pfam_df.merge(go_df)\n",
    "gene_annot_df.loc[:, 'gene_id'] = gene_annot_df.loc[:, 'transcript'].str.replace(\"Anogam_\", \"\").str.replace(\"-R[A-Z]\", \"\")\n",
    "\n",
    "# load normalised counts \n",
    "counts = pd.read_csv(normcounts_path, sep=\"\\t\")\n",
    "\n",
    "\n",
    "## define functions ## \n",
    "\n",
    "def plot_rotated_dendro(dendro, ax, linewidth=2):\n",
    "    coords = zip(dendro['dcoord'], dendro['icoord'])\n",
    "    lines = LineCollection([list(zip(x, y)) for x, y in coords], colors='lightgray', linewidth=linewidth)\n",
    "\n",
    "    ax.add_collection(lines)\n",
    "    number_of_leaves = len(order)\n",
    "    max_dependent_coord = max(map(max, dendro['dcoord']))\n",
    "\n",
    "    ax.yaxis.set_ticks_position('right')\n",
    "\n",
    "    # Constants 10 and 1.05 come from\n",
    "    # `scipy.cluster.hierarchy._plot_dendrogram`\n",
    "    ax.set_ylim(0, number_of_leaves * 10)\n",
    "    ax.set_xlim(0, max_dependent_coord * 1.05)\n",
    "\n",
    "    ax.invert_xaxis()\n",
    "    ax.invert_yaxis()\n",
    "\n",
    "    sns.despine(ax=ax, bottom=True, left=True, right=True)\n",
    "    ax.tick_params(top=False,\n",
    "                   bottom=False,\n",
    "                   left=False,\n",
    "                   right=False,\n",
    "                   labelleft=False,\n",
    "                   labelright=False,\n",
    "                   labelbottom=False)\n",
    "    return(ax)\n",
    "\n",
    "def gene_ids_from_domain(gene_annot_df, domain):\n",
    "    gene_list = []\n",
    "    if isinstance(domain, list):\n",
    "        for dom in domain:\n",
    "            ids = gene_annot_df.query(\"domain == @domain\")['gene_id']\n",
    "            gene_list.append(ids)\n",
    "            return(np.unique(gene_list))\n",
    "    else:\n",
    "        return(gene_annot_df.query(\"domain == @domain\")['gene_id'].to_numpy())\n",
    "\n",
    "\n",
    "### MAIN\n",
    "\n",
    "# a dict with gene families and their respective Pfam domain for extracting\n",
    "gene_fams = {'COE': 'COesterase', \n",
    "             'OBP':'PBP_GOBP', \n",
    "             'OR':'7tm_6', \n",
    "             'UGT':'UDPGT',\n",
    "             'Io':['Lig_chan','7tm_1'],\n",
    "             'Gr': '7tm_7',\n",
    "             'P450':'p450', \n",
    "             'Gst':['GST_N', 'GST_N_3', 'GST_C'], \n",
    "             'ABC':['ABC_membrane', 'ABC_tran'],\n",
    "            'FAS':'ketoacyl-synt',\n",
    "            'ELO':'ELO',\n",
    "            'FAD':'FA_desaturase',\n",
    "            'FAR':'NAD_binding_4'}\n",
    "\n",
    "\n",
    "### read in diff exp data\n",
    "genediff = {}\n",
    "counts_dict = {}\n",
    "for comp in comparisons['contrast']:\n",
    "    df = pd.read_csv(f\"results/genediff/{comp}.csv\", sep=\",\")\n",
    "    names = df[['GeneID', 'GeneName']]#, 'GeneDescription']]\n",
    "    genediff[comp] = df\n",
    "    \n",
    "    #loop through gene fams \n",
    "    fam_dict = {}\n",
    "    for fam, domain in gene_fams.items():\n",
    "        gene_ids = gene_ids_from_domain(gene_annot_df, domain)\n",
    "        fam_de = df.query(\"GeneID in @gene_ids\").copy()\n",
    "        \n",
    "        fam_de.loc[:, 'de'] = np.where(\n",
    "             np.logical_and(fam_de['padj'].between(0, 0.05, inclusive=\"both\"), fam_de['FC'].ge(1)), \n",
    "            'Upregulated', \n",
    "             np.where(\n",
    "                     np.logical_and(fam_de['padj'].between(0, 0.05, inclusive=\"both\"), fam_de['FC'].le(1)), 'Downregulated', 'No Diff. expression'\n",
    "             )\n",
    "        )\n",
    "        \n",
    "        fam_dict[fam] = fam_de\n",
    "        counts_dict[fam] = counts.query(\"GeneID in @gene_ids\")\n",
    "        counts_dict[fam] = counts_dict[fam].merge(names)\n",
    "        counts_dict[fam].loc[:, 'Label'] = [id_ + \" | \" + name if name != \"\" else id_ for id_, name in zip(counts_dict[fam]['GeneID'].fillna(\"\"), counts_dict[fam]['GeneName'].fillna(\"\"))]\n",
    "        \n",
    "    genediff[comp] = dict(fam_dict)\n",
    "\n",
    "colors = [\"#d9d9d9\", '#33cc33', '#cc66ff']\n",
    "pal = sns.color_palette(colors)\n",
    "\n",
    "def remap(x):\n",
    "    if x == 'Upregulated':\n",
    "        return(1)\n",
    "    elif x == 'Downregulated':\n",
    "        return(2)\n",
    "    elif x == 'No Diff. expression':\n",
    "        return(0)\n",
    "    else:\n",
    "        assert x == 'Nan', \"whats wrong with x\"\n",
    "    \n",
    "\n",
    "def check_fcs(diffexp, fam, comparisons):\n",
    "    \n",
    "    idx = np.random.choice(diffexp.shape[0])\n",
    "    gene = diffexp['GeneID'][idx]\n",
    "    \n",
    "    vals = diffexp.set_index('GeneID').query(\"GeneID == @gene\").to_numpy()[0]\n",
    "    \n",
    "    comp_vals = []\n",
    "    for comp in comparisons:\n",
    "        row = genediff[comp][fam].query(\"GeneID == @gene\")\n",
    "        if row['padj'].to_numpy() < 0.05:\n",
    "            if row['FC'].to_numpy() > 1:\n",
    "                comp_vals.append(1)\n",
    "            elif row['FC'].to_numpy() < 1:\n",
    "                comp_vals.append(2)\n",
    "        else:\n",
    "            comp_vals.append(0)\n",
    "    \n",
    "    assert all(np.array(comp_vals) == vals), \"The values do not match! ffs\"\n",
    "\n",
    "with PdfPages(\"results/genediff/GeneFamiliesHeatmap.pdf\") as pdf:\n",
    "\n",
    "    for fam, domain in gene_fams.items():\n",
    "\n",
    "        ids = counts_dict[fam]['GeneID'].to_numpy()\n",
    "        diffexp = counts_dict[fam]['GeneID'].to_frame().copy()\n",
    "\n",
    "        for comp in comparisons['contrast']:\n",
    "            de_data = genediff[comp][fam].set_index(\"GeneID\").reindex(ids).reset_index()\n",
    "            assert all(diffexp['GeneID'].to_numpy() == de_data['GeneID'].to_numpy()), \"wrong!!\"\n",
    "            diffexp.loc[:, comp] = de_data['de']\n",
    "\n",
    "        diffexp_remap = diffexp.set_index(\"GeneID\").reindex(index=ids)\n",
    "        diffexp_remap = diffexp_remap.applymap(remap)\n",
    "        #check_fcs(diffexp_remap.reset_index(), fam, comparisons['contrast'])\n",
    "\n",
    "        df = counts_dict[fam].set_index(\"Label\").drop(columns=['GeneID', 'GeneName'])\n",
    "        size = len(df)\n",
    "\n",
    "        g = sns.clustermap(data=np.log2(df+1), \n",
    "                        cmap=\"Blues\", \n",
    "                        vmax=10, \n",
    "                        cbar=False,\n",
    "                        linewidths=4,\n",
    "                        col_cluster=False,\n",
    "                        linecolor=\"white\",\n",
    "                        yticklabels=False, \n",
    "                        xticklabels=False,\n",
    "                        tree_kws=dict(linewidths=3, colors=\"lightgray\"),\n",
    "                        figsize=[1, 1])\n",
    "        ax = g.ax_heatmap\n",
    "        dendro = g.ax_row_dendrogram\n",
    "        g.ax_cbar.set_visible(False)\n",
    "        g.ax_row_dendrogram.set_visible(False) #suppress row dendrogram\n",
    "        g.ax_col_dendrogram.set_visible(False) #suppress column dendrogram\n",
    "\n",
    "        order = g.dendrogram_row.reordered_ind\n",
    "\n",
    "        diffexp_remap = diffexp_remap.iloc[order, :]\n",
    "        assert all(diffexp_remap.index == df.iloc[order, :].index.str[:10].to_list()), \"rows do not match\"\n",
    "\n",
    "        df = df.T.assign(sampleID=df.T.index)\n",
    "        df.loc[:, 'treatment'] = df['sampleID'].str.replace(\"[\\d+]\", \"\", regex=True)\n",
    "        n_treatments = len(df.loc[:, 'treatment'].unique())\n",
    "\n",
    "        fig, ax = plt.subplots(1, n_treatments+2, figsize=[10,(size/3)], gridspec_kw={'width_ratios':[3, 3] + list(np.repeat(1, n_treatments))})\n",
    "        fig.suptitle(f\"{fam}, pfam domains = {domain}\", fontsize=18, fontweight='bold')\n",
    "        dendro = scipy.cluster.hierarchy.dendrogram(g.dendrogram_row.linkage, no_plot=True,\n",
    "                                                        above_threshold_color='lightgray', \n",
    "                                                        color_threshold=0, \n",
    "                                                        no_labels=True)\n",
    "        ax[0] = plot_rotated_dendro(dendro, ax[0])    \n",
    "        sns.heatmap(ax=ax[1], data=diffexp_remap, cmap=pal, cbar=False, linewidths=4, linecolor='white')\n",
    "        ax[1].set_xticklabels(comparisons['contrast'], fontsize=14, fontweight='bold')\n",
    "        ax[1].set_ylabel(\"\")\n",
    "\n",
    "        for idx, group in enumerate(metadata.treatment.unique()):\n",
    "            idx += 2\n",
    "            df2 = df.query(\"treatment == @group\").drop(columns=['treatment', 'sampleID'])\n",
    "            df2 = df2.T.iloc[order, :]\n",
    "\n",
    "            sns.heatmap(ax=ax[idx], data=np.log10(df2+1), cmap=\"Blues\", cbar=False, linewidths=4, linecolor='white')\n",
    "\n",
    "            ax[idx].set_xlabel(group, rotation=90, fontsize=14, fontweight='bold', ha='right', rotation_mode='anchor')\n",
    "            ax[idx].set_xticklabels([])\n",
    "            ax[idx].set_ylabel(\"\")\n",
    "            ax[idx].tick_params(top=False,\n",
    "                                bottom=False,\n",
    "                                left=False,\n",
    "                                right=False,\n",
    "                                labelleft=False,\n",
    "                                labelright=True,\n",
    "                                labelbottom=True)\n",
    "\n",
    "            sns.despine(left=True, bottom=True, right=True)\n",
    "            ax[idx].set_yticklabels(labels=df2.index, rotation=0,fontsize=14)\n",
    "        for axes in ax[1:-1]:\n",
    "            axes.set_yticklabels([])\n",
    "            axes.set_yticks([])\n",
    "\n",
    "        fig.show()\n",
    "        fig.savefig(f\"results/genediff/{fam}_expression.tiff\", dpi=200)\n",
    "        pdf.savefig(bbox_inches = 'tight')"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
